### DISCUSSION TITLE
How does the ISnapper work?

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#1255] [Q&A]
Hi, 
I have a strange behavior since I migrated from GLSP 1.x to 2.x. In our [Open-BPMN project](https://github.com/imixs/open-bpmn/) we have some special cases where elements are forced to snap in specific positions. 

For example, a so called BPMN BoundaryEvent is always placed on the edge of a Task, and a Lane-Divider should not move out of its pool:

![Peek 2024-02-18 09-39](https://github.com/eclipse-glsp/glsp/assets/1339124/86b873fa-e3c0-48f5-a49b-627d832d8d19)

I solved this in the past with a custom `ISnapper` implementation, where I simply calculated the possible absolute new position of a snap-point. So the result of the method `snap(position: Point, element: SModelElement): Point`  was always an absolute position in the diagram plane. 

This is no longer true for GLSP version 2.x. I can't find any notes regarding the changed behavior in the release notes. I am very confused about this issue and I did not found a solution. This topic does not have any relation ship to the new HelperLine feature. 

If I run  my old code in GLSP 2.x it looks like this caused by the calculation of the absolute position:

![Peek 2024-02-18 09-54](https://github.com/eclipse-glsp/glsp/assets/1339124/4d780b8f-7893-4461-a164-d64e3621f6e5)

I wonder how the `ISnapper` is working and if using a  `ISnapper` may be the wrong approach for my feature to move some elements only aligned to a internal line... ?

**Comment by rsoika** [comment]
...maybe I should work with a `FeedbackCommand` instead of a snapper? I found that the helperline feature works with FeedbackCommands. 

And I found this [comment mentioning feedback commands](https://github.com/eclipsesource/graphical-lsp/issues/174#issuecomment-462349778)... But I did not found any code comments or docu. So maybe this is only for internal usage?

**Reply by martin-fleck-at** [reply]
I believe the behavior of the position calculation in some of the tools changed from working on the absolute position and handing that over to the snapper to having the snapper only snap the delta of the previous and current position. You are correct in that there seems to be no obvious mentioning of this somewhere and based on your previous feedback we already improved the [PR template](https://github.com/eclipse-glsp/glsp-client/blob/master/.github/PULL_REQUEST_TEMPLATE.md) so that such changes are better transferred and reported on release in the future.

If you need to snap absolute positions, you may need to override respective handler (e.g., this is how it was done in another project for the ghost element feedback: https://github.com/axonivy/process-editor-client/pull/475/files#diff-8c754a52dc5e2ac166d5b3bfb3653a990a22aec24043017b9f625525712a4133R86). In general, this is very valuable feedback to us so we know which parts of the framework may need better customizablity.

As for feedback commands, they are just regular commands with an additional priority and an already "empty" undo/redo handling, i.e., really more a convenience class that can be used to implement the handlers of feedback actions. All feedback actions that are registered with the feedback action dispatcher are automatically re-applied after each model update (in the order given by the priority of the handling command, see `FeedbackAwareUpdateModelCommand`). However, there is no need to use the `FeedbackCommand` base class as you can just use a regular command as well.

Client-side visual feedback such as moving an element (or creating an element with a ghost) is typically implemented as feedback that is active (i.e., registered with the feedback action dispatcher) while we are performing the action. So in your case, it might be that you need to adapt the position calculation in the `FeedbackMoveMouseListener` of the change-bounds tool or maybe you snapper implementation needs to be adapted.

**Reply by rsoika** [reply]
Hi @martin-fleck-at 
yes a `MouseTrackingElementPositionListener` sounds from it's name like something made to solve my requirement. But I do not understand how to adapt/overwrite this class, and where and how I should place my own code providing my computed new mouse position. Can you give me a short example? 

On the other hand, you mentioned that the snapper now only provides the delta of the  previous and current position. And this is exactly what I observed. Would you mean also a snapper can still be a good choice to recalculate the delta position in my case?

**Comment by rsoika** [comment]
To give you a better example of what I did with trying to adapt my snapper. 

I implemented (for example for the divider problem) the following snapper :


```javascript
@injectable()
export class BPMNElementSnapper implements ISnapper {
  constructor(public grid: { x: number; y: number } = { x: 1, y: 1 }) { }
  snap(position: Point, _element: GModelElement): Point {
    if (isLaneDivider(_element)) {
      return this.findLaneDividerSnapPoint(_element, position);
    }
    return {
      x: Math.round(position.x),
      y: Math.round(position.y)
    };
  }
  private findLaneDividerSnapPoint(element: GModelElement, position: Point): Point {
    const x = 0;
    const y = position.y;
    return { x: x, y: y };
  }
}
```

Very easy code. I just set the x(delta) always to 0.

And this looks at the first glance fine but if I move the mouse aggressive left/right the element suddenly changes the x position even when I set it to 0 - this is the strange thing.....

![Peek 2024-02-18 22-54](https://github.com/eclipse-glsp/glsp/assets/1339124/d0cd56cb-c120-4dd6-bc5d-2d732fe8c8f9)


This effect also happens when I move the mouse very fast.

**Reply by martin-fleck-at** [reply]
That is very interesting but it might have something to do with how the calculated position is used in the end. I think I might have to dig a little bit deeper into your code to see what is really going on. The state that you show above, is that in your current main branch or do you have a dedicated branch that I could have a look at?

**Comment by rsoika** [comment]
Hi @martin-fleck-at, I am sorry, but at the moment I did not understand anything. Even analyzing the code you referred did not help me. 

To me it looks over-killed extending the `FeedbackMoveMouseListener` or `MouseTrackingElementPositionListener`. At least I do not understand how to register such an extension in my own DiagramModule. Everything I tried so far breaks my modeller code completely with circular dependency exceptions and other strange errors. 

From the overall concept of GLSP I expect that I should be able to 

 - implement some kind of Action listener
 - compute my new element position depending on my internal diagram logic
 - send somehow a new action event to update the mouse position 

at least I understand that I can not or must not overwrite the bounds property of an element in any action as this is always readonly property - right?

Or did you expect that a `ISnapper` should still be a good place to adjust the mouse position? If so I will continue working on my iSnapper. 

You can see my current code base [here](https://github.com/imixs/open-bpmn/blob/76f990bef2f76f00ada6ad3cf6813308927fda99/open-bpmn.glsp-client/open-bpmn-glsp/src/di.config.ts#L91) and [here](https://github.com/imixs/open-bpmn/blob/76f990bef2f76f00ada6ad3cf6813308927fda99/open-bpmn.glsp-client/open-bpmn-glsp/src/bpmn-select-listeners.tsx#L50). But it is more experimental at the moment.

**Reply by martin-fleck-at** [reply]
I had a look at the code and I believe I know why the problem is happening: When you move the mouse really fast, the element that you are dragging (e.g., the lane divider) may not be moved fast enough and so the element we are trying to snap will be whatever else is beneath the mouse cursor. In most cases this will be the graph but it might be any other element as well (e.g., an edge).

Since we are calculating the position difference first (on whatever element), snap that and then apply the delta to all the moving elements, it can happen that your snapper gets a graph element and therefore returns a delta that allows for horizontal movement. You can see the whole calculation in the `FeedbackMoveMouseListener`: https://github.com/eclipse-glsp/glsp-client/blob/e7dec9bd52b9688a7a23005c3f7de652d0e85923/packages/client/src/features/tools/change-bounds/change-bounds-tool-feedback.ts#L193

So we can either adapt the FeedbackMoveMouseListener or - as a more slim workaround - adapt the snapper to not only consider the given element but also any selected elements, i.e., do something like this:

```ts
@injectable()
export class BPMNElementSnapper extends GridSnapper {
  constructor() {
    super({ x: 5, y: 5 });
  }

  override snap(position: Point, element: GModelElement): Point {
    if (isElementOrSelected(element, isLaneDivider)) {
      return { ...position, x: 0 };
    }

    // move routing-points by given grid
    if (isElementOrSelected(element, elem => 'volatile-routing-point' === elem.type)) {
      return super.snap(position, element);
    }

    // default move on a 1x1 grid
    return { x: Math.round(position.x), y: Math.round(position.y) };
  }
}

/**
 * Returns true if the given element or any of the selected elements in the indexed graph matches the given predicate.
 */
function isElementOrSelected(element: GModelElement, predicate: (modelElement: GModelElement) => boolean): boolean {
  try {
    return predicate(element) || getMatchingElements(element.index, isSelected).some(predicate);
  } catch (error) {
    return false;
  }
}
```
WARNING: I did not fully check the code so that might interfere with other places where the positioning is used.

**Reply by rsoika** [reply]
Hi @martin-fleck-at, your guess is absolutely correct. I've seen exactly this by log messages, that the graph element or other elements has actually been transferred. I will try your suggestion extending the `GridSnapper` and give you feedback asap.

**Reply by rsoika** [reply]
Yes - you code works and it looks like it is solving the problem! I will continue working with this and implement the other snapping rules....

**Reply by rsoika** [reply]
Hi @martin-fleck-at , I need to correct myself. It does not work. Or at least the method `isElementOrSelected` has no useful effect. 

From the following code snippet you would expect that `isElementOrSelected` is only true for type 'lane-divider'

```javascript
....
  override snap(position: Point, _element: GModelElement): Point {
    if (isElementOrSelected(_element, elem => 'lane-divider' === elem.type)) {
      console.log('  ---> element (' + _element.type + ')');
      return { ...position, x: 0 };
      //return this.findLaneDividerSnapPoint(_element, position);
    }

    // default move 1x1...
    return {
      x: Math.round(position.x),
      y: Math.round(position.y)
    };
  }
}
function isElementOrSelected(element: GModelElement, predicate: (modelElement: GModelElement) => boolean): boolean {
  try {
    return predicate(element) || getMatchingElements(element.index, isSelected).some(predicate);
  } catch (error) {
    return false;
  }
}
```

But the console.log shows also again the graph element :-(

![Peek 2024-02-19 22-34](https://github.com/eclipse-glsp/glsp/assets/1339124/ecadd23a-c403-4b9d-acfd-1ce0d42318c9)

**Reply by martin-fleck-at** [reply]
I think that the graph shows up is fine because you get the graph from the mouse event and then we go through the whole index of elements to find if a lane divider was selected. So the `element` is still the graph but somewhere along the lines a line divider is selected which is why this method returns `true`.

**Comment by rsoika** [comment]
Finally I found a working solution (not perfect). My code looks like this - Note that I need to explicit check the 'graph' case at the beginning of the snap method. Found no other solution. Only returning 0,0 in the 'graph' case healed the situation (very very strange):

```javascript
@injectable()
export class BPMNElementSnapper extends GridSnapper {
  constructor() {
    super({ x: 1, y: 1 });
  }

  override snap(position: Point, element: GModelElement): Point {
    if ('graph' === element.type) {
      console.log('causing problems...!');
      return { x: 0, y: 0 };
    }

    if (this.isElementOrSelected(element, elem => 'lane-divider' === elem.type)) {
      return this.findLaneDividerSnapPoint(element, position);
    }

    // default move 1x1...
    return {
      x: Math.round(position.x),
      y: Math.round(position.y)
    };
  }

  /**
   * Returns true if the given element or any of the selected elements in the indexed graph matches the given predicate.
   * 
   */
  public isElementOrSelected(element: GModelElement, predicate: (modelElement: GModelElement) => boolean): boolean {
    try {
      return predicate(element) || getMatchingElements(element.index, isSelected).some(predicate);
    } catch (error) {
      return false;
    }
  }

  /*
   * This helper method computes the snap Position of a Lane-Divider.
   * The position is based on the Bounds of the containing Pool.
   * The final position is always on the x position of the Pool.
   */
  public findLaneDividerSnapPoint(element: GModelElement, position: Point): Point {
    let x = 0;
    let y = position.y;
    // test min / max position
    if (hasArgs(element) && isBoundsAware(element)) {
      x = 0;
      const yMin = Number(element.args.ymin);
      const yMax = Number(element.args.ymax);

      console.log('  ---> element (' + element.type + ') position: ' + element.bounds.x + ',' + element.bounds.y + '   yMin/yMax=' + yMin + '/' + yMax
        + '  Point--> ' + position.x + ',' + position.y);

      if (element.bounds.y + position.y < yMin) {
        y = 0;
      }
      if (element.bounds.y + position.y > yMax) {
        y = 0;
      }
    }
    // return the new position;
    return { x: x, y: y };
  }
}
```
![Peek 2024-02-19 23-30](https://github.com/eclipse-glsp/glsp/assets/1339124/311fdde8-2ce0-4c10-9555-e29c41fb64f7)

**Reply by martin-fleck-at** [reply]
That is very curious indeed! I do not understand why you need to intercept the graph like this. If it is only the graph and nothing else is selected, the custom method should not trigger. Furthermore, if you return `0, 0` for the graph then we have the problem that when you move the mouse quickly and the snapper gets called for the graph `element` even though a lane divider is selected, there will be no movement, isn't that right?

**Reply by rsoika** [reply]
>  if you return 0, 0 for the graph.... when you move the mouse quickly ....... there will be no movement, isn't that right?

Yes

**Comment by rsoika** [comment]
Isn't it the `PointPositionUpdater` or `PositionSnapper` that are more relevant for my requirement? 
Isn't there a chance for me to register an additional custom BPMNPointPositionUpdater that is providing in some situations other decisions?

**Comment by rsoika** [comment]
Hi @martin-fleck-at , maybe we should take a step back and rethink the situation. I am afraid that we are now proceeding in a way that is far too complicated.
The main interface for diagram projects is the Action API. As a diagram developer, I should not adapt or change internal objects like the `GridSnapper` or `PositionSnapper`...

Why don't you just send in a method like  `PointPositionUpdater.updatePosition()`  an synchronous action. As a diagram, I can observe such an event (to recompute for example the position) or not. This action is only for diagram implementers and not for your own API development. So you separate such complex things from your own code. Is this possible or is sending a synchronous event to expensive in such a situation?

I'm doing the same thing in Jakarta EE in my [workflow engine](https://github.com/imixs/imixs-workflow/blob/1e81edbd3ec7e4d141f640ff7c737d3393e66c34/imixs-workflow-engine/src/main/java/org/imixs/workflow/engine/WorkflowService.java#L518). The engine transmits life cycle events at specific points in the process. These allow implementers to change the properties of a process instance. The engine itself doesn't use these events. It is just an interceptor interface.

**Reply by martin-fleck-at** [reply]
The `PointPositionUpdater.updatePosition` is definitely a good point to add adaptations in general. However, as far as I can see the `PointPositionUpdater` (currently not injectable) delegates to the `PositionSnapper` (injectable but does not receive the mouse event) and that one delegates to the `ISnapper`. Since our problem is that the element we get from the mouse event is not actually the element we are interested in (e.g., when moving we are interested in the element being moved independent from what the mouse event reports).

I think we could adapt the `PointPositionUpdater` and try to figure out which element we want by re-calculating the element at the given position but we would need to create a custom `FeedbackMoveMouseListener` and when we do that, we can already insert our custom behavior anyway. From a framework perspective, we should definitely improve this but for now - if the solution above - doesn't work for you, we might need to create our own custom change bounds tool with our own custom feedback listener (always based on the exiting ones of course).

**Reply by rsoika** [reply]
Ok, this sounds all not very easy to me. Can I implement such a `FeedbackMoveMouseListener` by myself or is this an api change you are thinking about? I can't follow your thoughts here. 

I can skip this topic a little bit as it is not the most important for me. I calculate the position on the server side anyway. This will look a little bit strange but it results in a valid model. 

What do you think about my idea to simply send an synchronous event that can be processed by an diagram implementer if needed.

**Reply by rsoika** [reply]
Hi @martin-fleck-at , sorry but I did not understand what I can do.
Extending the `FeedbackMoveMouseListener` did not work for me. Trying to bind my own `CustomFeedbackMoveMouseListener` leads to an Circular dependency. 
What did you mean with your last response?

 - should I extend  the `PointPositionUpdater`?
 - should I extend the `FeedbackMoveMouseListener`?
 - should I implement something like a `PositionSnapper`?
 - or should I just wait until you change some internal behavior? 

I think the requirement to fit a moved element to a specific position in a diagram could be a very general and often needed one. The easiest way for a diagram project would be to react on a specific event. If the event provides the ID of the moved element and the current move position, maybe it is possible that an eventListener can return a new custom position.

**Reply by martin-fleck-at** [reply]
Right, we would need to override the `FeedbackMoveMouseListener`. If you can wait until we make the core framework more extensible, that would be the simplest option. Alternatively, I can try to guide you through the steps that are necessary to implement the custom `FeedbackMoveMouseListener`.

**Reply by rsoika** [reply]
Ah OK - yes of course I can wait for you. No problem.

**Reply by martin-fleck-at** [reply]
Great, thank you! I'll try to get to that part soon as it is quite a core issue I'd say. If you need a solution before I can get to, please let me know and I'll try to guide you through the customization but basically what you would need to do is bind your own ChangeBoundsTool that inherits from the GLSP one and override the creation of the mouse listeners to provide your own custom mouse listener implementation with the adapted methods.

**Reply by rsoika** [reply]
Only if you like to have kind of prototype of your idea. Otherwise I can wait and test the next nightly build....
